### PR TITLE
clean heron java instance for stateful processing

### PR DESCRIPTION
**Pull request by nlu90**
During the testing of stateful processing with java HeronInstance, I found several problems:
1. Slave thread is destroyed during restore
2. logic of handling if the instance's slave looper should be started is confusing
3. IInstance interface is not very clean and well organized
4. stmgr accidently cleans instance info 

This PR is fixing the above mentioned issues and with some code refactoring. Tested locally now the stateful processing for java instance is working normally and even with failure cases like instance dying, stmgr dying, tmaster dying.

### CODE REVIEW

File: examples/src/java/com/twitter/heron/examples/StatefulWordCountTopology.java [hunk 1]
```diff
@@ -162,7 +168,13 @@ public void prepare(Map map, TopologyContext topologyContext, OutputCollector ou
 
     @Override
     public void execute(Tuple tuple) {
+      try {
+        Thread.sleep(3000);
+      } catch (Exception ex) {
+        // do nothing
+      }
       String key = tuple.getString(0);
+      System.out.println(key);
```
Comment (srkukarni) [inline examples/src/java/com/twitter/heron/examples/StatefulWordCountTopology.java]: Leftover debugging changes?

Comment (nlu90) [reply examples/src/java/com/twitter/heron/examples/StatefulWordCountTopology.java]: sure

File: heron/common/src/java/com/twitter/heron/common/utils/misc/PhysicalPlanHelper.java [hunk 2]
```diff
@@ -301,5 +302,15 @@ public boolean isTerminatedComponent() {
   public boolean isCustomGroupingEmpty() {
     return customGrouper.isCustomGroupingEmpty();
   }
+
+  public boolean isTopologyStateful() {
+    Map<String, Object> config = topologyContext.getTopologyConfig();
+    return String.valueOf(Config.TopologyReliabilityMode.EFFECTIVELY_ONCE).equals(
```
Comment (maosongfu) [inline heron/common/src/java/com/twitter/heron/common/utils/misc/PhysicalPlanHelper.java]: It is good to have a method to convert `config.get(Config.TOPOLOGY_RELIABILITY_MODE)` to enum and then compare two enum values. Comparing two Strings is error prone.

File: heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java [hunk 3]
```diff
@@ -135,17 +135,17 @@ public void persistState(String checkpointId) {
 
     // Checkpoint
     if (bolt instanceof IStatefulComponent) {
-      LOG.info("Starting checkpoint");
       ((IStatefulComponent) bolt).preSave(checkpointId);
     } else {
-      LOG.info("Trying to checkponit a non stateful component. Send empty state");
+      LOG.info("Trying to checkpoint a non stateful component. Send empty state");
```
Comment (maosongfu) [inline heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java]: Do we need the `LOG.info(...)` here?

File: heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java [hunk 4]
```diff
@@ -232,7 +235,7 @@ public void readTuplesAndExecute(Communicator<Message> inQueue) {
       if (msg instanceof CheckpointManager.InitiateStatefulCheckpoint) {
         String checkpointId =
             ((CheckpointManager.InitiateStatefulCheckpoint) msg).getCheckpointId();
-        LOG.info("Start checkpoint for: " + checkpointId);
+        LOG.info("Persisting state for checkpoint: " + checkpointId);
```
Comment (maosongfu) [inline heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java]: Do we need a `LOG.info(..)` here?

File: heron/instance/src/java/com/twitter/heron/instance/spout/SpoutInstance.java [hunk 5]
```diff
@@ -147,7 +147,6 @@ public void persistState(String checkpointId) {
     }
 
     if (spout instanceof IStatefulComponent) {
-      LOG.info("Starting checkpoint");
       ((IStatefulComponent) spout).preSave(checkpointId);
     } else {
       LOG.info("Trying to checkponit a non stateful component. Send empty state");
```
Comment (maosongfu) [inline heron/instance/src/java/com/twitter/heron/instance/spout/SpoutInstance.java]: Do we need a `LOG.info(..)` here?

Comment (nlu90) [reply heron/instance/src/java/com/twitter/heron/instance/spout/SpoutInstance.java]: removed this log

File: heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java (hunk 3 above)
Comment (nlu90) [reply heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java]: removed this log

File: heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java (hunk 4 above)
Comment (nlu90) [reply heron/instance/src/java/com/twitter/heron/instance/bolt/BoltInstance.java]: This log message is indicating the instance is doing checkpoint, we should keep it. But I moved this `LOG.info()` into `persistState` method for better organization

File: heron/common/src/java/com/twitter/heron/common/utils/misc/PhysicalPlanHelper.java (hunk 2 above)
Comment (nlu90) [reply heron/common/src/java/com/twitter/heron/common/utils/misc/PhysicalPlanHelper.java]: good suggestion. Fixed.

**Comment by maosongfu** [comment]
LGTM. 👍

**Pull request merged**